internal/appsec/waf/waf.go

```diff
@@ -223,73 +262,81 @@ type Context struct {
 // NewContext a new WAF context and increase the number of references to the WAF
 // handle. A nil value is returned when the WAF handle can no longer be used
 // or the WAF context couldn't be created.
-func NewContext(waf *Handle) *Context {
+func NewContext(handle *Handle) *Context {
 	// Increase the WAF handle usage count by RLock'ing the mutex.
 	// It will be RUnlock'ed in the Close method when the Context is released.
-	waf.mu.RLock()
-	if waf.handle == nil {
-		// The WAF handle got free'd by the time we got the lock
-		waf.mu.RUnlock()
+	handle.refCounter.RLock()
+	if handle.handle == nil {
+		// The WAF waf got free'd by the time we got the lock
+		handle.refCounter.RUnlock()
 		return nil
 	}
-	context := C.ddwaf_context_init(waf.handle)
+	context := C.ddwaf_context_init(handle.handle)
 	if context == nil {
 		return nil
 	}
 	incNbLiveCObjects()
 	return &Context{
-		waf:     waf,
+		handle:  handle,
 		context: context,
 	}
 }
 
 // Run the WAF with the given Go values and timeout.
-func (c *Context) Run(values map[string]interface{}, timeout time.Duration) (matches []byte, err error) {
+func (c *Context) Run(values map[string]interface{}, timeout time.Duration) (matches []byte, actions []string, err error) {
 	now := time.Now()
 	defer func() {
 		dt := time.Since(now)
 		c.totalOverallRuntimeNs.Add(uint64(dt.Nanoseconds()))
 	}()
-	return c.run(values, timeout)
-}
 
-func (c *Context) run(values map[string]interface{}, timeout time.Duration) (matches []byte, err error) {
 	if len(values) == 0 {
 		return
 	}
-	wafValue, err := c.waf.encoder.encode(values)
+
+	wafValue, err := c.handle.encoder.encode(values)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return c.runWAF(wafValue, timeout)
+
+	return c.run(wafValue, timeout)
 }
 
-func (c *Context) runWAF(data *wafObject, timeout time.Duration) (matches []byte, err error) {
+// run is the critical section of Run
+func (c *Context) run(data *wafObject, timeout time.Duration) (matches []byte, actions []string, err error) {
+	// Exclusively lock this WAF context for the time of this run
 	c.mu.Lock()
 	defer c.mu.Unlock()
+
+	// RLock the handle
+	c.handle.mu.RLock()
```

Comment:
This is in case a rule data update attempt is made at the same time, right?
May be worth it to explain it in the comment, took me a while to get it straight :sweat_smile:

Reply:
Adding a comment 👍 
The other way to read this is by literally reading the Go line of code "rlock of the mutex of the WAF handle" - meaning that you want read-access to the WAF handle.

Reply:
https://github.com/DataDog/dd-trace-go/pull/1493/commits/54ed9e5e36058ac08b99b9de4610839fa3cc53ce